### ISSUE TITLE
Environment.Exit hanging in RemoteExecutor

### ISSUE THREAD

**Issue by ViktorHofer** [dotnet/coreclr#27128]
https://dev.azure.com/dnceng/public/_build/results?buildId=380511&view=ms.vss-test-web.build-test-results-tab&runId=11763518&paneView=debug&resultId=147963

Configuration: `netcoreapp-Windows_NT-Debug-x64-Windows.81.Amd64.Open`

```
System.Net.Tests.AuthenticationManagerTest.CustomTargetNameDictionary_ValidCollection

Timed out at 10/7/2019 8:39:55 PM after 60000ms waiting for remote process.\r\n\tProcess ID: 5404\r\n\tHandle: 1096\r\n\tName: dotnet\r\n\tMainModule: C:\\dotnetbuild\\work\\ddad0c4a-3d2b-409b-b8b2-09e253a66230\\Payload\\dotnet.exe\r\n\tStartTime: 10/7/2019 8:38:55 PM\r\n\tTotalProcessorTime: 00:00:00.1875000\r\n\tThreads:\r\n\t\tThread dotnet/corefx#1 (OS 0x152C) \r\n\t\t\t[InlinedCallFrame] (System.Environment._Exit)\r\n\t\t\t[InlinedCallFrame] (System.Environment._Exit)\r\n\t\t\tSystem.Environment.Exit(Int32)\r\n\t\t\tMicrosoft.DotNet.RemoteExecutor.Program.Main(System.String[])\r\n\t\t\t[GCFrame]\r\n\t\t\t[GCFrame]\r\n\t\tThread dotnet/corefx#2 (OS 0x1538) [Finalizer]\r\n\t\t\t[DebuggerU2MCatchHandlerFrame]\r\n\t\tThread dotnet/corefx#3 (OS 0x1528) [Thread pool worker]\r\n
   at Microsoft.DotNet.RemoteExecutor.RemoteInvokeHandle.Dispose(Boolean disposing) in /_/src/Microsoft.DotNet.RemoteExecutor/src/RemoteInvokeHandle.cs:line 131
   at Microsoft.DotNet.RemoteExecutor.RemoteInvokeHandle.Dispose() in /_/src/Microsoft.DotNet.RemoteExecutor/src/RemoteInvokeHandle.cs:line 55
   at System.Net.Tests.AuthenticationManagerTest.CustomTargetNameDictionary_ValidCollection() in /_/src/System.Net.Requests/tests/AuthenticationManagerTest.cs:line 93
```

cc @davidsh, @wfurt, @scalablecory, @eiriktsarpalis, @karelz
cc @stephentoub for another hang

**Comment by jkotas** [comment]
@dotnet/dotnet-diag Could you please take a look?

This hang in EventPipe is one of the top causes of intermittent test failures and affects CI stability.

**Comment by jkotas** [comment]
Why do we need to call `CancelSynchronousIo` at all? What is not going to work if that call is deleted?

**Comment by josalem** [comment]
It looks like it was introduced in #25602 and modified in #25786.  If I recall, there was a small slowdown on the shutdown path from breaking out of the `Accept` call on the named pipe, so this was added to explicitly cancel the `Accept` on the server thread.  I took a quick look at the docs for that Windows API, and it sounds like it shouldn't be a blocking call, so I'm not sure off the top of my head what could be going on here.

**Comment by jkotas** [comment]
Trying to do cleanup on process shutdown is waste of time and a bug farm. You should only do the minimal cleanup during shutdown to cleanup machine-wide state. The process local state will be taken care of by the OS.

**Comment by sywhang** [comment]
I think we can safely remove this call, but that will regress https://github.com/dotnet/coreclr/issues/25463 - which was tracking a shutdown delay of 20-40ms showing up in PSCore's performance tests. A potential solution is to remove the call to `CancelSynchronousIo` and add an environment variable to disable diagnostics server from getting created in the first place for cases like the PSCore tests. Does that sound ok @jkotas?

/cc @sergiy-k who reported #25463 first.

**Comment by jkotas** [comment]
I do not understand why deleting the `CancelSynchronousIo` would introduce shutdown performance regression. If we delete this call, where do you expect the shutdown to be stuck for 20-40ms?

**Comment by sywhang** [comment]
If I recall correctly, I think what Jose said when he fixed the original issue by adding this call was that calling ::CloseHandle() on the diagnostics server thread was taking longer on some special hardware with hyper-v disabled because it was waiting for pending I/Os to complete. I didn't investigate the issue myself so I don't have the full context here, but I think we can remove it and come up with a better workaround for that particular case.

**Comment by jkotas** [comment]
> ::CloseHandle() on the diagnostics server thread 

You do not need to call CloseHandle on the diagnostic server thread during shutdown. You can leave the handle open and leave it to the OS to take care of it. 

Since the thread handle is not really needed for anything, you may consider closing it immediately after CreateThread returns it and not keeping it around at all - but that is orthogonal to the shutdown fix.

**Comment by jkotas** [comment]
Reopening for 3.1 port.

**Comment by sywhang** [comment]
@jkotas I was preparing a 3.1 port and realized that this actually doesn't apply for 3.1 because we never ported the "fix" to #25463 to 3.0 so the relevant code doesn't exist in the 3.1 branch.

**Comment by jkotas** [comment]
ok
